packages/apollo-engine-reporting/src/plugin.ts

```diff
@@ -0,0 +1,403 @@
+        // get our documentAST in the execution callback after it is parsed.
+        queryString = requestContext.source;
+
+        if (requestContext.request.variables) {
+          treeBuilder.trace.details = makeTraceDetails(
+            requestContext.request.variables,
+            options.sendVariableValues,
+            queryString,
+          );
+        }
+
+        const clientInfo = generateClientInfo(requestContext);
+        if (clientInfo) {
+          // While clientAddress could be a part of the protobuf, we'll ignore it for
+          // now, since the backend does not group by it and Engine frontend will not
+          // support it in the short term
+          const { clientName, clientVersion, clientReferenceId } = clientInfo;
+          // the backend makes the choice of mapping clientName => clientReferenceId if
+          // no custom reference id is provided
+          treeBuilder.trace.clientVersion = clientVersion || '';
+          treeBuilder.trace.clientReferenceId = clientReferenceId || '';
+          treeBuilder.trace.clientName = clientName || '';
+        }
+      }
+
+      let endDone: boolean = false;
+      function didEnd() {
+        if (endDone) return;
+        endDone = true;
+        treeBuilder.stopTiming();
+
+        treeBuilder.trace.fullQueryCacheHit = !!metrics.responseCacheHit;
+        treeBuilder.trace.forbiddenOperation = !!metrics.forbiddenOperation;
+        treeBuilder.trace.registeredOperation = !!metrics.registeredOperation;
+
+        // If the user did not explicitly specify an operation name (which we
+        // would have saved in `executionDidStart`), but the request pipeline made
+        // it far enough to figure out what the operation name must be and store
+        // it on requestContext.operationName, use that name.  (Note that this
+        // depends on the assumption that the RequestContext passed to
+        // requestDidStart, which does not yet have operationName, will be mutated
+        // to add operationName later.)
+        const operationName = requestContext.operationName || '';
+
+        // If this was a federated operation and we're the gateway, add the query plan
+        // to the trace.
+        if (metrics.queryPlanTrace) {
+          treeBuilder.trace.queryPlan = metrics.queryPlanTrace;
+        }
+
+        addTrace({
+          operationName,
+          queryHash: requestContext.queryHash!,
+          documentAST: requestContext.document,
+          queryString,
+          trace: treeBuilder.trace,
+          schemaHash: requestContext.schemaHash,
+        });
+      }
+
+      return {
+        parsingDidStart() {
+          ensurePreflight();
+        },
+
+        validationDidStart() {
+          ensurePreflight();
+        },
+
+        didResolveOperation() {
+          ensurePreflight();
+        },
+
+        executionDidStart() {
+          ensurePreflight();
+          return didEnd;
+        },
```

Comment:
I will add a comment that explains this, but as there are various phases in the request pipeline which can be bypassed (e.g. parsing and validation are skipped when we get a hit on the document store cache, execution is bypassed when hitting the full-response cache, etc.), I'm being particularly defensive here to ensure that we always kick things off.

This would be less necessary and possible to implement in `requestDidStart` if the actual precise location of the `requestDidStart` hooks wasn't _different_ between the `graphql-extensions` implementation and the plugin API.  Specifically, with the plugin API, we start `requestDidStart` _before_ APQ negotiation happens — which I think that makes more sense overall, but it means that we have a few other life-cycles that we need to kick within order to make sure we start _after_ APQ has been poked.

Reply:
No longer needing of commentary after b981b590328a6a321956ca851fb34d19f973aa2! 🎉